### PR TITLE
Parallelize PrivateDist scan

### PR DESCRIPTION
**Pull request by ronawho**
Optimize and parallelize scan for PrivateDist. The implementation just
gathers the PrivateArray into a local DefaultRectangular array, performs
the scan locally, and then scatters the results to the result array.
This is significantly faster than the default serial approach. It's 100x
faster at 512 nodes. PrivateDist scans are unlikely to be a performance
bottleneck for anybody but it was causing serialization warnings for
Arkouda.

Below are scalability numbers for PrivateDist scan. Results are a rate
(scans/sec) so higher is better. I'm just running:

```chpl
for 1..trials do
  var resEnds: [PrivateSpace] int = + scan numResults;
```

| nodes | before | now    |
| ----: | -----: | -----: |
|   1   | 56,400 | 44,000 |
|   2   |  8,200 |  7,500 |
|   4   |  3,700 |  5,000 |
|   8   |  2,000 |  3,800 |
|  16   |  1,200 |  2,800 |
|  32   |    640 |  2,200 |
|  64   |     90 |  1,750 |
| 128   |     40 |  1,400 |
| 256   |     15 |  1,000 |
| 512   |      6 |    600 |

There's a slight hit for single node since we're copying around a
little, but there's a 100x improvement at 512 nodes.

Resolves https://github.com/chapel-lang/chapel/issues/14840
Closes https://github.com/Cray/chapel-private/issues/870

### CODE REVIEW

**Comment by ronawho** [comment]
I have run testScanOps.chpl for:
 - std
 - memleaks
 - valgrind
 - baseline
 - gasnet
 - ugni

File: modules/dists/PrivateDist.chpl [hunk 1]
```diff
@@ -210,6 +210,24 @@ proc PrivateArr.dsiSerialWrite(x) {
...
 
+proc PrivateArr.doiScan(op, dom) where (rank == 1) &&
+                                  chpl__scanStateResTypesMatch(op) {
+  type resType = op.generate().type;
+  var res: [dom] resType;
+
+  var localArr: [0..numLocales-1] resType;
+
+  coforall loc in Locales do on loc do
+    localArr[here.id] = if _local then data else chpl_getPrivatizedCopy(this.type, this.pid).data;
```
Comment (ronawho) [inline modules/dists/PrivateDist.chpl]: This is the line I'm least sure of, but it passes testing in all the config I tried. When accesses externally, ChapelArray's `proc this` takes care of getting privatization right, but when accessing the data internally here, `dsiAccess` doesn't give you a privatized copy so I'm doing it manually here. The part that's not obvious to me is when exactly privatization is enabled. It seems to be disabled for `--local` compilations, but there's also a compiler flag and some other conditions that make me less sure this is always right.

Comment (benharsh) [reply modules/dists/PrivateDist.chpl]: I think you'd technically want to mimic what ``_isPrivatized`` from ChapelArray does and check the magic ``_privatization`` value. In practice I don't know of a case when we've ever disabled privatization, so if time is a factor I'd be OK with this going in as-is.

Comment (ronawho) [reply modules/dists/PrivateDist.chpl]: Think it's worth switching to using `_isPrivatized` directly. I think that's just:

```chpl
localArr[here.id] = if _isPrivatized(this) then chpl_getPrivatizedCopy(this.type, this.pid).data else data;
```

Comment (ronawho) [reply modules/dists/PrivateDist.chpl]: (FYI `--baseline` disables privatization, but it looks like even hello4-datapar-dist.chpl is broken with `--comm=gasnet --baseline` on master, so maybe I'm worried about nothing here.)

**Comment by cassella** [comment]
Would it be worth having an `if numLocales == 1` case with the previous behavior?

**Comment by ronawho** [comment]
Yeah, considered that, but single node is still fast enough even with the regression here, so it didn't seem worthwhile to special case.

**Pull request merged**